pkg/sql/opt/invertedidx/geo.go

```diff
@@ -61,6 +90,129 @@ func IsGeoIndexFunction(fn opt.ScalarExpr) bool {
+// first two arguments are swapped.
+//
+// Returns the original function if commuteArgs is false, or a new function
+// representing the same relationship but with commuted arguments if
+// commuteArgs is true. For example:
+//
+//   ST_Intersects(g1, g2) <-> ST_Intersects(g2, g1)
+//   ST_Covers(g1, g2) <-> ST_CoveredBy(g2, g1)
+//
+// See commuteGeoRelationshipMap for the full list of mappings.
+//
+// Also returns the column ID corresponding to the input geospatial column
+// as well as the column ID corresponding to the inverted index key column,
+// which are the two columns relevant for the join condition. Returns ok=true
+// if a join condition was successfully extracted or ok=false otherwise.
+//
+// The given function must be a valid geospatial function that can be index-
+// accelerated, and must contain two variable arguments. Otherwise,
+// TryGetInvertedJoinCondFromGeoFunc will panic.
+func TryGetInvertedJoinCondFromGeoFunc(
+	factory *norm.Factory, fn opt.ScalarExpr, commuteArgs bool, inputProps *props.Relational,
+) (_ opt.ScalarExpr, inputGeoCol, indexGeoCol opt.ColumnID, ok bool) {
+	if !IsGeoIndexFunction(fn) {
+		panic(errors.AssertionFailedf(
+			"TryGetInvertedJoinCondFromGeoFunc called on a function that cannot be index-accelerated",
+		))
+	}
+	function := fn.(*memo.FunctionExpr)
+
+	// Extract the the variable inputs to the geospatial function.
+	if function.Args.ChildCount() < 2 {
+		panic(errors.AssertionFailedf(
+			"all index-accelerated geospatial functions should have at least two arguments",
+		))
+	}
+
+	arg1, arg2 := function.Args.Child(0), function.Args.Child(1)
+	if commuteArgs {
+		arg1, arg2 = arg2, arg1
+	}
+
+	// The first argument should come from the input.
+	variable, ok := arg1.(*memo.VariableExpr)
+	if !ok {
+		panic(errors.AssertionFailedf(
+			"TryGetInvertedJoinCondFromGeoFunc called on function containing non-variable inputs",
+		))
+	}
+	if !inputProps.OutputCols.Contains(variable.Col) {
+		return nil, 0, 0, false
+	}
+	inputGeoCol = variable.Col
+
+	// The second argument should be a variable corresponding to the index
+	// column. The caller must verify that this column is actually the key column
+	// of an inverted geospatial index.
+	variable, ok = arg2.(*memo.VariableExpr)
+	if !ok {
+		panic(errors.AssertionFailedf(
+			"TryGetInvertedJoinCondFromGeoFunc called on function containing non-variable inputs",
+		))
+	}
+	indexGeoCol = variable.Col
+
+	if commuteArgs {
+		// Copy the original arguments into a new list, and swap the first two
+		// arguments.
+		args := make(memo.ScalarListExpr, len(function.Args))
+		copy(args, function.Args)
+		args[0], args[1] = args[1], args[0]
+
+		// Get the geospatial relationship that is equivalent to this one with the
+		// arguments commuted, and construct a new function that represents that
+		// relationship.
+		rel := geoRelationshipMap[function.Name]
+		commutedRel := commuteGeoRelationshipMap[rel]
```

Comment:
What happens if a function has IsGeoIndexFunction == true, but doesn't have an entry in commuteGeoRelationshipMap? It might occur that there are geo functions that can be index accelerated but also can never be argument commuted? I suggest checking for `, ok` here to ensure that commutedRel is not the zero value.